dataflow/src/main/java/org/checkerframework/dataflow/cfg/AbstractCFGVisualizer.java

```diff
@@ -0,0 +1,337 @@
+                    if (exception.startsWith("java.lang.")) {
+                        exception = exception.replace("java.lang.", "");
+                    }
+
+                    for (Block b : blocks) {
+                        sbDigraph.append(addEdge(cur.getId(), b.getId(), exception));
+                        if (!visited.contains(b)) {
+                            visited.add(b);
+                            workList.add(b);
+                        }
+                    }
+                }
+            }
+
+            cur = workList.poll();
+        }
+
+        sbDigraph.append(generateNodes(visited, cfg, analysis));
+
+        // footer
+        sbDigraph.append("}\n");
+
+        return sbDigraph.toString();
+    }
+
+    protected abstract String generateNodes(
+            Set<Block> visited, ControlFlowGraph cfg, @Nullable Analysis<A, S, T> analysis);
+
+    protected String addEdge(long sId, long eId, String labelContent) {
+        return "    " + sId + " -> " + eId + " [label=\"" + labelContent + "\"];\n";
+    }
+
+    /**
+     * @param cbFooter Footer for conditional block
+     * @param osFooter Footer for the other situations
+     */
+    protected String visualizeBlockHelper(
+            Block bb,
+            @Nullable Analysis<A, S, T> analysis,
+            String cbFooter,
+            String osFooter,
+            String escapeCharacter) {
+        StringBuilder sbBlock = new StringBuilder();
+        sbBlock.append(loopOverBlockContents(bb, analysis, escapeCharacter));
+
+        // handle case where no contents are present
+        boolean centered = false;
+        if (sbBlock.length() == 0) {
+            centered = true;
+            if (bb.getType() == Block.BlockType.SPECIAL_BLOCK) {
+                sbBlock.append(visualizeSpecialBlock((SpecialBlock) bb));
+            } else if (bb.getType() == Block.BlockType.CONDITIONAL_BLOCK) {
+                sbBlock.append(cbFooter);
+                return sbBlock.toString();
+            } else {
+                sbBlock.append(osFooter);
+                return sbBlock.toString();
+            }
+        }
+
+        // visualize transfer input if necessary
+        if (analysis != null) {
+            // the transfer input before this block is added before the block content
+            sbBlock.insert(0, visualizeBlockTransferInput(bb, analysis));
+            if (verbose) {
+                Node lastNode = getLastNode(bb);
+                if (lastNode != null) {
+                    StringBuilder sbStore = new StringBuilder();
+                    sbStore.append(escapeCharacter).append("~~~~~~~~~").append(escapeCharacter);
+                    sbStore.append("After:");
+                    sbStore.append(visualizeStore(analysis.getResult().getStoreAfter(lastNode)));
+                    sbBlock.append(sbStore);
+                }
+            }
+        }
+
+        sbBlock.append((centered ? "" : escapeCharacter));
```

Comment:
This is easier to read as:

```java
if (!centered) {
    sbBlock.append(escapeCharacter);
}
```

Maybe use a variable that contains the negation instead, to not require the negation here.